lib/r10k/environment/tarball.rb

```diff
@@ -0,0 +1,78 @@
+require 'r10k/util/setopts'
+require 'r10k/tarball'
+require 'r10k/environment'
+
+class R10K::Environment::Tarball < R10K::Environment::WithModules
+
+  R10K::Environment.register(:tarball, self)
+
+  # @!attribute [r] tarball
+  #   @api private
+  #   @return [R10K::Tarball]
+  attr_reader :tarball
+
+  include R10K::Util::Setopts
+
+  # Initialize the given tarball environment.
+  #
+  # @param name [String] The unique name describing this environment.
+  # @param basedir [String] The base directory where this environment will be created.
+  # @param dirname [String] The directory name for this environment.
+  # @param options [Hash] An additional set of options for this environment.
+  #
+  # @param options [String] :source Where to get the tarball from
+  # @param options [String] :version The sha256 digest of the tarball
+  def initialize(name, basedir, dirname, options = {})
+    super
+    setopts(options, {
+      # Standard option interface
+      :type      => ::R10K::Util::Setopts::Ignore,
+      :source    => :self,
+      :version   => :checksum,
+
+      # Type-specific options
+      :checksum => :self,
+    })
+
+    @tarball = R10K::Tarball.new(name, @source, checksum: @checksum)
+  end
+
+  def path
+    @path ||= Pathname.new(File.join(@basedir, @dirname))
+  end
+
+  def sync
+    tarball.get unless tarball.cache_valid?
+    case status
+    when :absent, :mismatched
+      tarball.unpack(path.to_s)
+      # Untracked files left behind from previous extractions are expected to
+      # be deleted by r10k's purge facility.
+    end
+  end
+
+  def status
+    if not path.exist?
+      :absent
+    elsif not (tarball.cache_valid? && tarball.insync?(path.to_s, ignore_untracked_files: true))
+      :mismatched
+    else
+      :insync
+    end
+  end
+
+  def signature
+    @checksum || @tarball.cache_checksum
+  end
+
+  include R10K::Util::Purgeable
+
+  # Returns an array of the full paths to all the content being managed.
+  # @note This implements a required method for the Purgeable mixin
+  # @return [Array<String>]
+  def desired_contents
+    desired = []
+    desired += @tarball.paths.map { |entry| File.join(@full_path, entry) }
```

Comment:
Discovered an interesting edge case here: if you build the tarball a bit wrong, even though it unpacks correctly, it might purge incorrectly. In my case, I created the tarball with `.`. So the paths inside it look like this:
```
"/Users/aileen/code/r10k/yaml-envs/tarball/.",
 "/Users/aileen/code/r10k/yaml-envs/tarball/./site-modules",
 "/Users/aileen/code/r10k/yaml-envs/tarball/./hiera.yaml",
...
```
But r10k's `current_contents` list looks like this:
```
"/Users/aileen/code/r10k/yaml-envs/tarball/site-modules",
 "/Users/aileen/code/r10k/yaml-envs/tarball/hiera.yaml",
...
```
So it purges everything except the `modules` directory (which is ignored by the `environment` purge level).

I realize this was my mistake, and I should get better at tarballs 😛 But it seems like something that could trip people up and cause a very confusing bug. Maybe we can add some normalization in `Tarball.paths`? Adding `Pathname.new(path).cleanpath` seemed to do the trick for me.

Reply:
Ha! Excellent catch. Although I do have unbridled confidence that users of software would NEVER imperfectly construct a tarball 😉, this does seem like something we could fix.

Reply:
Huh. I just snap-pushed what I thought would be the fix in 5a1711e, and then realized I was looking at a completely different part of the code. But, it might actually be a valid fix anyway? Don't have a test for it yet.

Reply:
That's the spot I edited when fixing this locally 🤷‍♀️ Seemed to work for my bad tarball.

Reply:
Were you working on a test for this? I think besides that, I'm happy to merge this.

Reply:
Yeah, lemme add some thing for that. I might just re-construct the fixture tarball to include the `./` name components, I think that'd make it easy to validate. Will push something shortly.

Reply:
Yeah that's what I was thinking 🙂